### ISSUE TITLE
New post Cake: expressive algorithms and data structures in Scala

### ISSUE THREAD

**Issue by postmillenial** [postmillenial/reading-list#638]
Cake: expressive algorithms and data structures in Scala<br>
http://ift.tt/2fX0uVV<br><div class="kg-card-markdown">
<p><em>Cake</em> is a Scala library for expressing algorithms and data structures in a declarative and typed manner, focusing on properties like:</p>
<ol>
<li>structure (directed graph, hierarchy)</li>
<li>propagation of information (signal, feedback, random flow)</li>
<li>parametrized node types</li>
<li>IO behavior</li>
<li>storage implementation (adjacency matrix/list)</li>
</ol>
<p>It utilizes the cake pattern in Scala, combining traits in as flexible a manner as possible, helped via statically-checked dependency hierarchies of the mentioned properties.</p>
<ul>
<li>Cake repo: <a href="http://ift.tt/2ybUrGZ">http://ift.tt/2g6PSYj</a>
</li>
<li>backpropagation example used in this article: <a href="http://ift.tt/2yc0mfe">Backprop.scala</a>
</li>
</ul>
<h1 id="backpropagationwithcake">Backpropagation with Cake</h1>
<p>A good example of the kind of algorithm and data structure pairing that can be created with Cake is an implementation of backpropagation for supervised learning in neural networks.</p>
<p>The relevant structure here is a <strong>hierarchical graph</strong>, i.e. a set of layers of <strong>nodes</strong> and <strong>directed weighted connections</strong> between them, with a beginning input layer, N hidden layers, and a final output layer. The operations which the algorithm utilizes include propagating a <strong>signal</strong> from the input layer to the output layer, propagating the error signal (<strong>feedback</strong>) in the opposite direction (backpropagation), and finally using the accumulated error information to update the weight values of the connections (an isolated process/<strong>flow</strong>, not a directed signal).</p>
<p>I'll explain the code for the backpropagation algorithm in Cake in a fragmented and nonlinear manner, focusing on the most important aspects first.</p>
<h2 id="essentialdefinition">Essential definition</h2>
<pre>
<code class="language-scala">val bp = new Nodes[State with Order]
        with Hierarchy
        with Directed with AdjacencyList
        with Signal
        with Feedback
        with Flow
        with IO {
</code>
</pre>
<p>This piece of code begins the definition of the <code>bp</code> object--containing the necessary data with the right structure and supported operations for a backpropagation algorithm.</p>
<ol>
<li>
<p><code>Nodes[State with Order]</code><br>
This trait specifies that the bp object is a collection of nodes of type <code>State with Order</code>. <code>State</code> and <code>Order</code> are some basic characteristics Cake provides by default for nodes: the former adds a simple <code>state: Double</code> to nodes, and the latter is utilized to order the nodes inside a hierarchy.</p>
</li>
<li>
<p><code>Hierarchy</code><br>
Tags the structure as a hierarchical ordering of nodes (e.g. a neural network or a binary tree). It specifies a requirement on the <code>Nodes</code> trait that it be parametrized <code>with Order</code>, a trait that adds an <code>order: Int</code> property to nodes.</p>
</li>
</ol>
<p><img src="http://ift.tt/2g7Zq5j" alt="fig_neural_network_1"></p>
<ol start="3">
<li>
<p><code>Directed with AdjacencyList</code><br>
The <code>Directed</code> trait adds weighted one-directional connections between pairs of nodes. Practically, each connection is a <code>Double</code> value signifying its weight/strength. However, they can either be existent or nonexistent, so the implementation wraps them as an <code>Option[Double]</code>, and the methods for accessing and modifying them respect that interface (e.g. <code>bp(1, 2) = None</code> removes a connection from node 1 to node 2).</p>
<p><code>AdjacencyList</code> specifies neither a structural nor an operational property: it's merely an implementational detail that functions with the <code>Directed</code> trait and stores the weights in an adjacency list. This line of code could for example be changed to <code>Directed with AdjacencyMatrix</code> so that the actual storage for values in the structure is implemented in a different manner--with no other changes to the code required.</p>
</li>
<li>
<p><code>Signal</code>, <code>Feedback</code> and <code>Flow</code><br>
In addition to specifying structural properties, it is also useful to specify operational properties that relate to how the structures processes information, or in a more user-oriented lexicon provide basic operations necessary to implement more complex algorithms. Obviously, structure and operation are codependent, and Cake ensures that this dependence is satisfied through Scala's type system. (Currently the only structures of interest are graphs, so these three traits specifically describe the flow of information through a directed, weighted graph.)</p>
<ul>
<li>
<p><code>Flow</code> signifies that information flows through the structure without any structural or temporal dependence--a free-for-all via all connections. It applies to any structure that is the subtype of <code>Nodes[_] with Directed</code>, meaning that practically any graph can support it--however, this is still a statically-checked requirement.</p>
</li>
<li>
<p><code>Signal</code> and <code>Feedback</code> are more specific types of information flow, they signify flows that have a structural dependence. In practice this means that the structure has to be hierarchical--a neural network is a great example, as it is essentially not an unordered set of nodes, but a set of layers (1 input, N hidden, 1 output layer most commonly). These two traits require that the <code>Hierarchy</code> trait precede them. <code>Hierarchy</code>, in turn, specifies that the nodes are subtypes of <code>Order</code>, a type which <code>Signal</code> and <code>Feedback</code> operate upon.</p>
</li>
</ul>
</li>
<li>
<p><code>IO</code><br>
Another mostly implementational concern, but crucial nonetheless as it provides protocols for processing inputs and providing outputs from the structure, and the protocol enables connecting to other IO providers and processors.</p>
<p>In the case of backpropagation, the IO provider (or source) could be a collection of pairs of input and target vectors, while the IO processor (or sink) could report on learning epochs (or even control the learning process itself).</p>
</li>
</ol>
<h2 id="nodes">Nodes</h2>
<p>Cake doesn't have a specific node type that other nodes extends, rather, nodes are merely assemblages of various arbitrary properties. One example is the attribute of having a state, or a label signifying the node is used as an output.</p>
<p>All nodes in a neural network obviously need a <code>State</code> attribute and an <code>Order</code> attribute (as they exist in a hierarchy). Those in the first layer need <code>Input</code>, and those in the last the <code>Output</code> label. Additionally for backpropagation, the output nodes will also contain a <code>Target</code> attribute, all nodes except input ones need an <code>Error</code> attribute, and there exists a single instance of a special <code>Bias</code> node in each non-output layer.</p>
<p>This is the code for these three node attributes:</p>
<pre>
<code class="language-scala">trait Error {
    this: State =&gt;
    var error: Double = 0.0
}

trait Target extends Error {
    this: Output =&gt;
    var target: Double = 0.0
}

trait Bias extends Input {
    state = 1
}
</code>
</pre>
<p>Utilization of the cake pattern is evident here as well. In the <code>Error</code> trait definition, a constraint is placed using the <code>this: State =&gt;</code> self-type Scala syntax, so that the trait always has to be combined with a <code>State</code> node (as nodes with no states cannot err). Further on, the <code>Target</code> trait extends <code>Error</code>, placing a constraint that only <code>Output</code> nodes may have the <code>target: Double</code> property. Finally, <code>Bias</code> nodes are input nodes with their states set to 1.</p>
<p>Nodes are <strong>accessed</strong> through the <code>nodes</code> field/method: <code>bp.nodes(nodeIndex)</code>.</p>
<h2 id="initializingbpwithnodes">Initializing <code>bp</code> with nodes</h2>
<p>Inside the definition of the <code>bp</code> object,</p>
<pre>
<code class="language-scala">// val bp = ...
//          ...
//          ... {

    val nodes =
        Hierarchy(2, 2, 1).nodes( // non-bias nodes
            () =&gt; new {val order = 0} with Input with Order,
            () =&gt; new {val order = 1} with State with Order with Error,
            () =&gt; new {val order = 2} with Output with Order with Target
        ) ++
        Nodes.nodes( // bias nodes
            ((() =&gt; new {val order = 0} with Bias with Order), 1),
            ((() =&gt; new {val order = 1} with Bias with Order), 1)
        )
</code>
</pre>
<p>The <code>nodes</code> property is an abstract member of the <code>Nodes[T]</code> trait, and is simply whichever covariant sequence of T: <code>def nodes: Seq[_ &lt;: T]</code>. This abstract property obviously has to be defined by the implementing object, and here it's a simple <code>val</code> created with two builder methods <code>Hierarchy(rows...).nodes</code> and <code>Nodes.nodes</code>. The hierarchy helper transforms a set of instructions for creating a node in layer i to N layers with M nodes.</p>
<p>The usage of the factory functions is obvious from their types:</p>
<h3 id="hierarchyordered">Hierarchy (ordered):</h3>
<pre>
<code class="language-scala">def nodes[T &lt;: Order : TypeTag](createInLayer: (()=&gt;T)*)
</code>
</pre>
<p>The <code>TypeTag</code> implicit parameter is irrelevant for functionality, what matters is that this function creates nodes of type <code>T</code> which is a subtype of <code>Order</code>. Another example of a constraint being placed: <code>createInLayer</code> is a sequence of builder functions that can produce any type of nodes as long as they mix in the <code>Order</code> trait.</p>
<p>What's hidden here is the <code>Hierarchy(2, 2, 1)</code> part of the code in the example: this is a simple <code>apply(Int*)</code> function on the companion object <code>Hierarchy</code> that determines the dimensions of each layer (in this case, 2 nodes in the first and second layer, 1 in the final one--again, that's input/hidden, .../output). That information is utilized by the factory function to call the relevant builder methods the correct number of times.</p>
<h3 id="nodesunordered">Nodes (unordered):</h3>
<pre>
<code class="language-scala">def nodes[T : TypeTag](fs: (()=&gt;T, Int)*)
</code>
</pre>
<p>This function takes pairs of builder functions and natural numbers that determine how many nodes each function should produce. It's obviously similar to the ordered factory function, with the distinction that no type constraint is placed upon the node objects produced.</p>
<h2 id="populatingthestructurewithweights">Populating the structure with weights</h2>
<p>The last section explained the process of initializing the structure with nodes, however, the connections between them are yet to be defined. Connecting the nodes is always done outside of the definition of an object, however utility methods for that are provided as well.</p>
<p>The code <code>bp &lt;= Hierarchy(2, 2, 1).random(-1, 1)</code> will initialize the <code>bp</code> object's weights to random real numbers between -1 and 1: however this only applies to nodes created with the <code>Hierarchy.nodes</code> helper method, which means that bias nodes are excluded. The operator <code>&lt;=</code> is provided by the <code>Directed</code> trait, and the <code>Hierarchy.random(a, b)</code> method returns an initializer object expected as an argument by this operator.</p>
<p>Since the bias nodes are specific to this algorithm, they're set manually with the following code:</p>
<pre>
<code class="language-scala">bp.of[Bias with Order]. // all nodes from bp that are subtypes of Bias with Node
map(in =&gt; (in._1 -&gt; in._2.order)). // (index, node) -&gt; (index, order)
foreach { case (i: Int, inRow: Int) =&gt;
    bp.order(inRow+1). // all nodes in the layer above bias
    filter(!bp.nodes(_).isInstanceOf[Bias]). // that aren't bias nodes themselves
    foreach { j =&gt; bp(i, j) = Some(Util.random(-1, 1)) } // connect from bias (i) to node (j)
}
</code>
</pre>
<p>The <code>of[T &lt;: N]</code> method is provided by <code>Nodes[N]</code> trait and returns all of the nodes with the specific subtype requested mapped from their index number in the collection of nodes. The results of this method are usually then mapped, filtered, and processed in other ways for specific operations upon the relevant nodes' properties.</p>
<h3 id="accessingweights">Accessing weights</h3>
<p>The <code>Directed</code> trait adds two apply methods that are used to access the weights with either <code>this(...)</code> inside the object definition or <code>bp(...)</code> outside.</p>
<ol>
<li>
<p><code>this(i: Int, j: Int): Option[Double]</code>: connection from node i to node j, <code>None</code> if nonexistent, <code>Some[Double]</code> if existent</p>
</li>
<li>
<p><code>this(i: Int): Map[Int, Double]</code>: the map of connections <em><strong>to</strong></em> node i from various other indices, i.e. keys for the map (nonexistent connections not included)</p>
</li>
</ol>
<p>A third method <code>from(i: Int): Map[Int, Double]</code> returns an analogous map of outward connections from node i.</p>
<h2 id="definingiooperations">Defining I/O operations</h2>
<p>The <code>IO</code> trait requires the two following methods to be implemented:</p>
<pre>
<code class="language-scala">def input(channels: Map[String, Seq[Double]]): Unit
def output(): Map[String, Seq[Double]]
</code>
</pre>
<p>The types of these functions are revealing: information is received and transmitted as sequences of doubles mapped onto channels, identified by names (strings).</p>
<p>The input for the backpropagation implementation are input values and the target output values, so these sequences are provided on appropriately-named channels <code>"inputValues"</code> and <code>"targetValues"</code>. Its output essentially doesn't matter during the training, however the network still exposes the relevant states on channels <code>"inputs"</code>, <code>"outputs"</code>, and <code>"errors"</code>.</p>
<p>This is the implementation of the above two functions by the backpropagation object:</p>
<pre>
<code class="language-scala">def input(channels: Map[String, Seq[Double]]) = {
    (channels("inputValues") zip of[Input].values).
    foreach { case (x: Double, i: Input) =&gt; i.state = x }

    (channels("targetValues") zip of[Target].values).
    foreach { case (x: Double, t: Target) =&gt; t.target = x }

    signal(propagate)
    feedback(backpropagate)
    flow(updateWeight)
}
</code>
</pre>
<p>The first part of the input function is again self-descriptive: it assigns input and target values to the relevant nodes. The second part is more interesting: after recording input, the method also processes it. The three function calls (<code>signal</code>, <code>feedback</code>, and <code>flow</code>) form the backbone of the backpropagation algorithm. The functions that do the actual heavy-lifting--<code>propagate</code>, <code>backpropagate</code>, and <code>updateWeight</code>--will be explained later.</p>
<p>The output method:</p>
<pre>
<code class="language-scala">def output = Map(
    "inputs" -&gt; of[Input].values.toSeq.map(_.state),
    "outputs" -&gt; of[Output].values.toSeq.map(_.state),
    "errors" -&gt; of[Error].values.toSeq.map(_.error)
)
</code>
</pre>
<p>This is extremely simple, a simple map is created mapping channel names to node values at the time of the call.</p>
<h3 id="operators">Operators</h3>
<p>The IO trait is rather useless by itself, the point however is in the operators for chaining I/O operations it provides:</p>
<ol>
<li>
<code>&gt;&gt;(to: IO)</code>: connects the left-hand side IO object's outputs to the right-hand-side IO object's input.</li>
<li>
<code>+(b: IO)</code>: combines the input and output methods of the two provided IO objects into a new IO object</li>
</ol>
<h2 id="thealgorithm">The algorithm</h2>
<p>As noted previously the actual backprop is implemented through three helper functions, which are passed as parameters to signal, feedback, and flow methods. These helper functions have a very restrictive type: <code>Int=&gt;Unit</code>, which implies that the helper function will be applied to a sequence of node indices. The ordering of this sequence differentiates the background methods signal, feedback, and flow, in a manner previously elaborated.</p>
<p>For propagation:</p>
<pre>
<code class="language-scala">def propagate(i: Int) = nodes(i).state =
    sig(this(i) map { case (i: Int, weight: Double) =&gt; nodes(i).state * weight } sum)
</code>
</pre>
<p>This is the simplest of the three helper functions, and its operation is obvious to anyone familiar with neural networks: the logistics/sigmoid function applied to the sum of input and weight products from the previous layer.</p>
<p>Backpropagating the error:</p>
<pre>
<code class="language-scala">def backpropagate(i: Int) = 
    nodes(i).asInstanceOf[Error].error = nodes(i) match {
        case o: Output with Target =&gt; (o.state - o.target)

        case s: State with Error =&gt; {
            val weights = from(i)

            (weights map { edge =&gt;
                weights(edge._1) * nodes(edge._1).asInstanceOf[Error].error * dsig(nodes(edge._1).state)
            } sum)
        }
    }
</code>
</pre>
<p>The only side-effect of the <code>backpropagate(i: Int)</code> function is setting the <code>error</code> property of the i-th node. The code simply expects that the node at the index i will be of type <code>Error</code>.</p>
<p>The <code>from(i: Int)</code> function was mentioned before as returning a map of indices to nodes that the node with index i connects to; here, it is utilized to backpropagate the error.</p>
<p>Finally, the <code>flow</code> method is called with the following function to update the weights:</p>
<pre>
<code class="language-scala">def updateWeight(i: Int) = this(i) = this(i) map { case (from: Int, weight: Double) =&gt;
    from -&gt; (weight - learningRate * nodes(i).asInstanceOf[Error].error * dsig(nodes(i).state) * nodes(from).state)
}
</code>
</pre>
<p>The weight update could have been achieved via either <code>signal</code> or <code>feedback</code> methods, however the <code>flow</code> method is used to avoid unnecessary sorting of nodes: weight deltas once the error has been calculated are independent of each other in this algorithm.</p>
<h2 id="puttingitalltogether">Putting it all together</h2>
<p>Once the algorithm and the structure have been defined, a set of input and target values is defined, and an IO chain is constructed:</p>
<pre>
<code class="language-scala">val inputs = List(List(0.0, 0.0),
                  List(0.0, 1.0),
                  List(1.0, 0.0),
                  List(1.0, 1.0))

val or = List(0.0, 1.0, 1.0, 1.0) map (x=&gt;List(x))
val xor = List(0.0, 1.0, 1.0, 0.0) map (x=&gt;List(x))

def pickRandom(i: Int) = scala.util.Random.nextInt(inputs.length)

val training = Revolve("inputValues" -&gt; inputs,
                       "targetValues" -&gt; xor)(pickRandom) &gt;&gt;
               bp &gt;&gt;
               Print(10000)
</code>
</pre>
<p>Most of the above code is rather self-explanatory. <code>Revolve</code> is an IO operator which projects randomly-picked, mutually-dependent value sequences to the desired output channels.</p>
<h2 id="runningtheexample">Running the example</h2>
<p>Since the <code>bp</code> object is an IO element, it is natural to call the <code>output</code> method on it. The piping constructed via the <code>&gt;&gt;</code> operator ensures that the input/target values, processing, and reporting are done in a correct sequential order. The integer argument to the <code>Print</code> IO sink (10000) specifies that a report should be printed each 10000 iterations (epoch size).</p>
</div>
<br><br>
via flowing &rarr; systems http://flowing.systems/<br>
October 09, 2017 at 05:35PM
